easybuild/tools/toolchain/toolchain.py

```diff
@@ -1049,39 +1060,70 @@ def handle_sysroot(self):
+        dependency_roots = []
+        for dep in dependencies:
+            if dep.get("external_module", False):
                 # for software names provided via external modules, install prefix may be unknown
-                names = dep['external_module_metadata'].get('name', [])
-                roots.extend([root for root in self.get_software_root(names) if root is not None])
+                names = dep["external_module_metadata"].get("name", [])
+                dependency_roots.extend([root for root in self.get_software_root(names) if root is not None])
             else:
-                roots.extend(self.get_software_root(dep['name']))
+                dependency_roots.extend(self.get_software_root(dep["name"]))
 
-        for root in roots:
-            self.variables.append_subdirs("CPPFLAGS", root, subdirs=cpp_paths)
-            self.variables.append_subdirs("LDFLAGS", root, subdirs=ld_paths)
+        for root in dependency_roots:
+            self._add_dependency_cpp_headers(root, extra_dirs=cpp)
+            self._add_dependency_linker_paths(root, extra_dirs=ld)
+
+    def _add_dependency_cpp_headers(self, dep_root, extra_dirs=None):
+        """
+        Append prepocessor paths for given dependency root directory
+        """
+        header_dirs = ["include"]
+        if isinstance(extra_dirs, (tuple, list)):
+            header_dirs.extend(extra_dirs)
+        header_dirs = set(header_dirs)  # remove duplicates
+
+        # mode of operation is defined by search-path-cpp-headers option
+        # toolchain option has precedence over build option
+        cpp_headers_mode = DEFAULT_SEARCH_PATH_CPP_HEADERS
+        tc_opt = self.options.option("search-path-cpp-headers")
+        build_opt = build_option("search_path_cpp_headers")
+        if tc_opt:
+            cpp_headers_mode = tc_opt
+            self.log.debug("search-path-cpp-headers set by toolchain option: %s", cpp_headers_mode)
+        elif build_opt:
+            cpp_headers_mode = build_opt
+            self.log.debug("search-path-cpp-headers set by build option: %s", cpp_headers_mode)
+
+        if cpp_headers_mode not in SEARCH_PATH_CPP_HEADERS:
+            raise EasyBuildError(
+                "Unknown value selected for option search-path-cpp-headers: %s. Choose one of: %s",
+                cpp_headers_mode, ", ".join(SEARCH_PATH_CPP_HEADERS)
+            )
+
+        for env_var in SEARCH_PATH_CPP_HEADERS[cpp_headers_mode]:
+            self.log.debug("Adding header paths to toolchain variable '%s': %s", env_var, dep_root)
+            self.variables.append_subdirs(env_var, dep_root, subdirs=header_dirs)
+
+    def _add_dependency_linker_paths(self, dep_root, extra_dirs=None):
+        """
+        Append linker paths for given dependency root directory
+        """
+        lib_dirs = ["lib64", "lib"]
+        if isinstance(extra_dirs, (tuple, list)):
+            lib_dirs.extend(extra_dirs)
+        lib_dirs = set(lib_dirs)  # remove duplicates
```

Comment:
Make it a set from the start? Also what happens if it is not an iterable or None? Maybe just assume it is and let Python throw if it isn't
```suggestion
        lib_dirs = {"lib64", "lib"}
        if extra_dirs:
            lib_dirs.update(extra_dirs)
```

Reply:
Thanks for the comment, this made realize that we should keep the order of those dirs as that translates into the order of paths in the corresponding environment variable. So I replaced going the set by `tools.utilities.nub`.

Regarding the `if instance`, I don't see a way to avoid it, because if a string is passed to this method that will be successfully taken by `extend` or `update` (as it is an iterable) but generating the wrong result.

Reply:
Good point, I added extra checks to handle unexpected types as `extra_dirs`. See https://github.com/easybuilders/easybuild-framework/pull/4645/commits/cd01b6564498618f689937af9bdd7904e56a1e73